### DISCUSSION TITLE
Talos API behind loadbalancer

### DISCUSSION THREAD

**Discussion by dekavox** [siderolabs/talos#8886] [Q&A] [answered]
Hi everybody,

I have a talos cluster running on hetzner (1 controlplane, 1 worker, 1 external loadbalancer) for test purposes.
Each of my nodes has a public and a private ip address (public ips and dns were obfuscated in the following examples) and all components are connected to each other via a private network. The nodes have firewalls configured but no open ingress ports by default.

**controlplane1**
public ip: 123.234.0.2
private ip: 10.0.1.11

**worker1**
public ip: 123.234.0.3
private ip: 10.0.1.101

**loadbalancer1**
public dns: "my-talos-cluster.mydomain.com"
public ip: 123.234.0.1
private ip: 10.0.1.2

The loadbalancer forwards ports 6443 and 50000 to the controlplane node and I added the public ip (and dns) of the loadbalancer as an endpoint in my talosconfig. For the nodes I defined the private ips in the talosconfig. Additionally I added the public ip and dns of the loadbalancer as a `.machine.certSANs` to my `controlplane.yaml` to ensure that these are added to the talos api certificate.

Unfortunaltely, if i run 
```bash
talosctl --talosconfig talosconfig apply-config --insecure --nodes 10.0.1.11 --file controlplane.yaml
```
it seems like talosctl is trying to directly communicate with the node via the private ip (instead of connecting via the endpoint). The same happens if I explicitly set the endpoint ip or dns.
```bash
talosctl --talosconfig talosconfig apply-config --insecure --endpoints 123.234.0.1 --nodes 10.0.1.11 --file controlplane.yaml
```
or
```bash
talosctl --talosconfig talosconfig apply-config --insecure --endpoints my-talos-cluster.mydomain.com --nodes 10.0.1.11 --file controlplane.yaml
```

Obviously this is not going to work, as my client has no access to the nodes via their private ip addresses.

As fas as I understand [this](https://www.talos.dev/v1.7/introduction/prodnotes/#load-balancing-the-talos-api) documentation, the endpoint should act as some kind of gateway/reverse-proxy to the nodes for the talos api, or am I getting it wrong?

I only get it to work if I open port 50000 on my nodes firewall and add the nodes via their public ip to my talosconfig (so connecting directly to them).

Is there a way how I can  use my loadbalancer as a single entrypoint for the talos api as well, keeping my firewall ports closed? The documentation above implies that this should be possible. What am I getting wrong?

**Comment by smira** [comment]
Endpoint is what `talosctl` connects to. `nodes` is what is used to communicate within the cluster (request reaches the endpoint and then gets forwarded within the cluster).

Your issue is about `apply-config --insecure` which is different in that regard, you should use the IP of your machine as it is reachable directly, `apply-config --insecure` can't work via other node as a router, so the request should go directly from your machine to the machine you're trying to reach.

You need to have direct connectivity to the machine for `apply-config --inscure`, either on IP layer or via a load-balancer of some sort. Or you can use other methods to supply machine config to the node (userdata, `talos.config=` for bare-metal, etc.).

**Reply by dekavox** [reply]
Hi @smira,

thanks for the fast response! 
As far as I understand it, the "reverse-proxy" function for the talos api is available only after configuration of the node(s)?

Do all cluster nodes have to be configured or is it enough to configure one controlplane node (via the loadbalancer or directly) and can I configure the worker nodes afterwards using the first controlplane node as an endpoint?
In this case I at least wouldn't have to open the ports for the worker nodes in the firewall.

Otherwise the automatic provisioning via ´talos.config=´ is a good alternative hint ;-)

Cheers
Dennis

**Reply by smira** [reply, marked as answer]
All nodes need to be fully configured before you can use endpoint/nodes. The thing is that Talos API uses mutual TLS (both for client-server and machine-to-machine communication). As the machine has no configuration, it has no PKI, so there's no way it could do mutual TLS.

So this `apply-config --insecure` is very special in that regard.

There are many other things you could do, e.g. you could use Omni which manages that part behind the SideroLink tunnel.

Or, you could build your own solution on top of SideroLink.

You could use `talos.config=` kernel arg to make machines pull the machine config over the network from some location. 

There are many ways here, certainly using Omni is the easiest path.

**Reply by dekavox** [reply]
Ok, thanks for the clarification!
I'll checkout the suggested options.

Best regards
